### DISCUSSION TITLE
[Feature Request] get /search result as JSON or query

### DISCUSSION THREAD

**Discussion by mahasadhu** [Laravel-Backpack/community-forum#315] [General]
# Feature Request

### What's the feature you think Backpack should have?

First of all I'm not sure whether this is not possible or already implemented, sorry if it is. 

Can we somehow mimic the datatable's request to the `/search` method ?
But instead of returning json with datatable-compatible format, we simply get the bare json based on the `$this->crud->getEntries()` ?

Or even better, we could try to find a way to create a method / route which can be used as a callback, that gets the same entries as the one currently shown in the datatable. 
That callback could receive either query or the eloquent collection.

This functionality could be useful to customize how we want to show the filtered data, for example for custom export layout.

### Have you already implemented a prototype solution, for your own project?

No I have not, but I think we can start by modularizing the `search()` method in ListOperation.php

### Do you see this as a core feature or an add-on?

Not sure

**Comment by tabacitu** [comment]
Hi @mahasadhu ,

It should be super-easy to [create a custom operation](https://backpackforlaravel.com/docs/4.1/crud-operations#creating-a-custom-operation) starting from ListOperation's `search()` method, but that returns stuff in a non-datatable-format, but a regular Eloquent collection JSON. Just copy-paste the code from `ListOperation::search()`

One thing I didn't understand from your post, sorry, was _how you'll be using this response_ (what you need this operation for). Is it to consume the API on the front-end of the application? If so, my plan was similar but in a slightly different direction, in using the [JSON-API](https://jsonapi.org) standard. That would have the benefit of providing a clear, documented and non-language standard of communication (the URL itself). 

The problem is - there's not much related to Backpack in creating these JSON-API routes. It's not Backpack-specific, it's just Laravel-specific, and AFAIK they can already be created pretty easily using something like https://github.com/cloudcreativity/laravel-json-api . Of course, we could make it _easier_, by creating a Backpack operation for it.

Let me know if I misunderstood your plan on how to use the API itself. I'll move this to Discussions so that it's easier to gather feedback and make plans, before we implement the specific feature.

Cheers!

**Reply by mahasadhu** [reply]
Thank you for your reply @tabacitu 

Sorry looks like I was not clear enough. What I meant by **mimic the datatable's request to the /search method** is to call the search method with the exact same URL params (GET parameter) and form data (POST parameter) as the one that datatable do in it's ajax request.

Because as we know, datatable send it's own POST form data to the /search endpoint containing it's ordering, search query, page, etc. Meanwhile backpack filter is modifying the URL with GET query parameters.
And I'm not sure how to get and combine those params into a single request, still tinkering with it.

Basically I need to get the "state" of the current list shown to the user, with all of its filter, sort, search query, page, etc

_how I'll be using this response_:
Yes, I need this response in a front end application, but I can also use it in backend, for example send the current list shown to someone's email, custom export / print layout, etc.
I was thinking to just add `?json=1` or any other flag to the URL if I need it to return json instead of eloquent collection or view.

But nevertheless, that custom operation is a good insight! Sorry I didn't know that functionality exists. 
Previously I just made a function inside the controller and add it to the route. Will redo it using the custom operation.

**Reply by tabacitu** [reply]
Oh... I get it now! Well if you create a custom operation, you could easily create a special route for it, so instead of `/search` it could be something like `/json`. Exactly the same, just not the same first segment.

**Comment by jcastroa87** [comment]
Hi.

Due to not activity on many years i will close this discussion, if there is needed feel free to reopen or create a new one.

Cheers.
